### ISSUE TITLE
Regression of 0.6.9, multiprocessing code crashing on some complex calls

### ISSUE THREAD

**Issue by deajan** [Nuitka/Nuitka#877]
Been tracking that bug for a while.
When importing module `psutil`, even without calling it, programs that worked with `multiprocessing` produce windows errors 
With nuitka => 0.6.9.1, I get a lot of windows errors (see below the kind of error I get).
The subprocesses spawned by multiprocessing die, and are respawned over and over again in an endless loop.

This behavior has been tested with Python 3.7 and Python 3.8, on nuitka 0.6.8.4 (working), nuitka 0.6.9.1 (not working), nuitka 0.6.9.6 ('not working) and 06.10.rc7 (not working).

Windows error (from event log):
```
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
- <System>
  <Provider Name="Application Error" /> 
  <EventID Qualifiers="0">1000</EventID> 
  <Version>0</Version> 
  <Level>2</Level> 
  <Task>100</Task> 
  <Opcode>0</Opcode> 
  <Keywords>0x80000000000000</Keywords> 
  <TimeCreated SystemTime="2020-11-08T18:44:49.2396924Z" /> 
  <EventRecordID>39269</EventRecordID> 
  <Correlation /> 
  <Execution ProcessID="0" ThreadID="0" /> 
  <Channel>Application</Channel> 
  <Computer>NPEDGE-RUNNER</Computer> 
  <Security /> 
  </System>
- <EventData>
  <Data>nuitka_mp_test.exe</Data> 
  <Data>0.0.0.0</Data> 
  <Data>5fa8370a</Data> 
  <Data>python38.dll</Data> 
  <Data>3.8.6150.1013</Data> 
  <Data>5f6b6fe5</Data> 
  <Data>c0000005</Data> 
  <Data>00000000000402a2</Data> 
  <Data>5628</Data> 
  <Data>01d6b5ff3c82a2a7</Data> 
  <Data>C:\GIT\PYTHON_BUILD64\nuitka_mp_test.dist\nuitka_mp_test.exe</Data> 
  <Data>C:\GIT\PYTHON_BUILD64\nuitka_mp_test.dist\python38.dll</Data> 
  <Data>85458ad4-805c-4260-8f3a-d3329085a641</Data> 
  <Data /> 
  <Data /> 
  </EventData>
  </Event>
```

* Nuitka version, full Python version and Platform (Windows, OSX, Linux ...)

0.6.10rc7
Python: 3.8.6 (tags/v3.8.6:db45529, Sep 23 2020, 15:52:53) [MSC v.1927 64 bit (AMD64)]
Executable: c:\Python38-64\python.exe
OS: Windows (10 20H2)
Arch: x86_64

* How did you install Nuitka and Python (pip, anaconda, deb, rpm, from source,
  what is a virtualenv ...), this is very important usually.

Installed via pip everytime.

* If possible please supply a [Short, Self Contained, Correct, Example](http://sscce.org/)
  that demonstrates the issue i.e a small piece of code which reproduces
  the issue and can be run with out any other (or as few as possible)
  external dependencies.

Here's the SSCCE, works when commenting out `import psutil`

```
import multiprocessing
import logging
from datetime import datetime, timedelta
from time import sleep
import psutil


logger = logging.getLogger()
logger.addHandler(logging.StreamHandler())
logger.setLevel(logging.DEBUG)


restrict_numbers = lambda n, n_min, n_max: max(min(n_max, n), n_min)


def dummy(arg):
    print('dummy n° {}'.format(arg))


def mp_runner(timeout=30):
    child_list = {}
    try:
        # Let's use half of the CPUs available but no more than 6 simultaneous WMI requests
        processes = int(multiprocessing.cpu_count() / 2)
        processes = restrict_numbers(processes, 1, 6)
        logger.debug('Running with %s processes.' % processes)
        mp_pool = multiprocessing.Pool(processes=processes)
        for i in range(1,12):
            child_list[i] = mp_pool.apply_async(dummy, [i])

        # Let the childs a chance to finish in 5 minutes
        childs_have_finished = False
        finish_time = datetime.now() + timedelta(seconds=timeout)
        while datetime.now() <= finish_time and childs_have_finished is False:
            childs_have_finished = True
            for name, child in child_list.items():
                if child.ready() is False:
                    childs_have_finished = False
            sleep(1)

        # Make sure we get output from every child that has finished, otherwise unregister it in list
        for name, child in child_list.items():
            if child.ready() is False:
                logger.warning('Timeout ! Ignoring [%s] query.' % name)
                child_list[name] = []

            else:
                # noinspection PyBroadException
                try:
                    child_list[name] = child.get(timeout=10)
                except Exception:
                    logger.warning('Cannot get query result.' % name)
                    logger.info('Trace:', exc_info=True)

        mp_pool.terminate()
    except (TimeoutError, multiprocessing.context.TimeoutError):
        logger.error('Multiprocessing timeout reached. Results will be incomplete.')
    except Exception:
        logger.error('Unhandled multiprocessing exception.')
        logger.error('Trace:', exc_info=True)
    return child_list

if __name__ == '__main__':
    mp_runner()
```

* If this is a regression (used to work in an earlier version of Nuitka),
  please note what you know about that.

Tried with multiple nuitka versions. Last version that worked was 0.6.8.4.

**Comment by deajan** [comment]
Seems related to #867

**Comment by kayhayen** [comment]
Thanks for this fine reproducer, going to find it now. :)

**Comment by kayhayen** [comment]
It just worked for me, I believe this might be fixed by avoiding the `PyObject_HasAttr` in Nuitka apparently, or at least it doesn't trigger anymore, or would be doing it with proper error. Right now I believe though, this is running into timeouts, as the forking for the multiprocessing module must rely on `sys.executable` which got changed first in 0.6.9.1 and since has been improved, but never in a way to restore the ability to use it.

**Comment by kayhayen** [comment]
I was trying to look into the `sys.executable` situation, only to find out, that we do the tricks in the code of the multiprocessing module, to set `sys.executable` to the binary path, and we have been doing that ever since. So my suspicions are not warranted.

**Comment by kayhayen** [comment]
So, this is reproducing it for me: `Assertion failed: (attr_name) != NULL, file C:\Users\kayha\repos\Py2C\nuitka\build\static_src\HelpersAttributes.c, line 502` with a twist that my `HAS_ATTR_BOOL` didn't keep the promise to not leak exceptions, it does in the factory version at least.

**Comment by deajan** [comment]
I've tested 0.6.9 tag today, it does trigger the error too.

**Comment by kayhayen** [comment]
That's very suprising, as this is about pre-made constants being used only. Your 0.6.9.0 confirmation indeed points to the commit you bi-sected. Somehow the value it tries to look at might not be successfully made. This is a first of that kind error, I was pretty confident the constant creation code is near perfect. But that's code for you, never is. Let's see the life cycle of the constant in play here, and which one it is.

**Comment by kayhayen** [comment]
It's just sad, that I never leveled debuggers for Windows. :/

**Comment by kayhayen** [comment]
The whole point of my factory change, was to make the attribute error lookup raise proper assertions from Nuitka over a segfault, so this is now more debuggable. But it's supposedly impossible to happen still. Except that I have made a bug there. The code that is called to register the meta path based freezer, runs before the global constants are assigned.... :/

**Comment by kayhayen** [comment]
Now I have to wonder, why this ever worked for anything at all.

**Comment by kayhayen** [comment]
When tracing, the module init trace for module level constants, gets called only once for every module. It seems to me, it should happen more often, and that would be inline with them not being initialized. But it could also just be buggy complex call helper re-formalation, I need to check.

**Comment by kayhayen** [comment]
Ok, I think I got this understood. We are doing something wrong for multiprocessing, since the `__internal__` module got removed. Its code now lives in `__main__` and that's always used first, right? Wrong! For multiprocessing, `__parents_main__` is used, with just the module name changed, so that `if __name__ == `__main__` can do the right things.

**Comment by kayhayen** [comment]
But for multiprocessing, we not call `__main__` ever, but code that does complex calls and happens to need a constant for doing it, e.g. to check `__iter__` attribute... well it's going to find it uninitialized, on Windows only, with multiprocessing only.

**Comment by kayhayen** [comment]
I need to divise a workaround, that e.g. will load these constants of the `__main__` even if it's never imported, or duplicate the internal helper functions, or revive the `__internal__` module again. So this is understood now.

**Comment by deajan** [comment]
I just don't get it why everything works okay when psutil module isn't loaded.

**Comment by kayhayen** [comment]
The psutil code uses a complex call helper that lives in `__main__` and that one uses a constant value that is not assigned. Nuitka does have a function set that works with `f(a, e=1, *b, **d)` and merges them to tuples and dicts, as those as the only ones, handling various cases, mappings vs. dicts, tuples different from, others.... and that makes it appear here.

**Comment by kayhayen** [comment]
I could be adding the constants used in the helper code to the global default constant population, that will resolve the issue, because if they only use global stuff, then they cannot fail. But I am trying to keep that set minimal, as they bloat every program. 

And I could come with a call to `__main__` in `__parents_main__` that initializes its constants too. That will require a bit of hand waiving, with `static` and `extern`, i.e. local vs. global C functions, but it should be workable.

**Comment by kayhayen** [comment]
Nice, this under control now, will soon surface on factory, simply calling that module constants init, will do the trick.

**Comment by deajan** [comment]
I'll happily recompile my project with latest factory to validate the fix ;)

**Comment by kayhayen** [comment]
It's now there, give it a spin. If you are really daring, try `--onefile` too. But that fix is going to appear as a hotfix very soon. Not sure if I should attempt to rebase master for it, maybe over the edge with that, and I am going to make a 0.6.10 soon anyway.

**Comment by deajan** [comment]
I tried latest factory, on my huge project, looks like it works great ;)
I'll have a full compilation now (32 and 64 bit) and have everything tested on a couple of different platforms.

As for the --onefile option, it didn't succeed yet. I'll open a new issue in order to not polluate this one.

Nuitka fixes on sunday... That's cool ;)

**Comment by kayhayen** [comment]
Once my buildbots have seen some of it, I am going to make that hotfix, family excusion coming up, but generally I work on Nuitka Fr-Mo. inclusive, and typically my day of it rather the first one of these, but then I tend to sprint.

**Comment by kayhayen** [comment]
This is part of 0.6.9.7 hotfix release.
